Add CellAssign ridgeplot (#437)
AlexsLemonade/scpca-nf: templates/qc_report/celltypes_qc.rmd

```

```{r, eval = has_cellassign}
knitr::asis_output("
### `CellAssign` assessment

> `CellAssign` computes the probability that each cell is one of the provided cell types and ultimately annotates cells based on the highest probability.

Reviewer: ```suggestion
`CellAssign` computes the probability that each cell in the library is one of the provided cell types in the reference and ultimately annotates cells by assigning the cell type with the highest probability.
```
I don't know if I did much better, but I tried to clarify between reference cell types and the cell types in our dataset?

We therefore expect that cell-level probabilities that correspond to the annotated cell types will be high.

> Conversely, we expect that cell-level probabilities that correspond to annotated cell types will be low.

Reviewer: ```suggestion
Conversely, we expect that cell-level probabilities that correspond to unannotated cell types will be low.
```

Author: good catch, thanks!

In the plot below, we show distributions of the `CellAssign` probabilities for each assigned cell type (excluding unknown cell types).

> Colors represent whether the probabilities correspond to a cell with that given annotation: Purple distributions represent probabilities for cells that _were not_ annotated as the given cell type.
> Yellow distribution represent probabilies for cells that _were_ annotated as the given cell type.

Reviewer: ```suggestion
Colors represent whether the probabilities correspond to a cell with that given annotation: Yellow distribution represent probabilies for cells that _were_ annotated as the given cell type.
Purple distributions represent probabilities for cells that _were not_ annotated as the given cell type.
```
This is minor and I don't know why my brain thinks this way, but I feel like it makes sense to mention the annotated cells first before unannotated.

Author: > but I feel like it makes sense to mention the annotated cells first before unannotated.

you are definitely right!

For distributions with fewer than three points, density plots cannot be calculated.
In these cases, we directly show values for individual cell probabilities as points.
")
```

```{r, eval = has_cellassign, warning=FALSE, message=FALSE,fig.height = 7, fig.width = 8}

# Prepare CellAssign scores for plot
cellassign_prob_df <- metadata(processed_sce)$cellassign_predictions |> 
  tidyr::pivot_longer(
    -barcode, 
    names_to = "celltype", 
    values_to = "probability"
  ) |>
  # remove cell types that were not annotated
  # bonus: ths will also remove the "other" level, which we don't want to show anyways

>   dplyr::filter(celltype %in% celltype_df$cellassign_celltype_annotation) |>

Reviewer: I actually think we do want to label the Unknown cell types here? This is different than `SingleR` where cells that are not assigned are just thrown out. Here `other` is an actual label where `CellAssign` is calculating a probability. I think it would be helpful to show the probabilities for cells that weren't assigned to labels vs. cells that were assigned to labels?

new_levels <- levels(cellassign_prob_df$cellassign_celltype_annotation)
new_levels <- rev(new_levels[-length(new_levels)]) # reverse for ridgeplot
cellassign_prob_df$annotation_wrapped <- factor(
  cellassign_prob_df$celltype,
  levels = new_levels,
  labels = stringr::str_wrap(new_levels, 30)
)

# find groups with <=2 cells to add back into plot
# These will always be `annotated = TRUE`
celltypes_leq2_cells <- cellassign_prob_df |>
  dplyr::count(annotation_wrapped, annotated) |>
  dplyr::filter(n <= 2) |>
  dplyr::pull(annotation_wrapped)

# data frame for adding those points back
leq2_probabilities_df <- cellassign_prob_df |>
  dplyr::filter(
    annotation_wrapped %in% celltypes_leq2_cells, 
    annotated
  ) |>
  dplyr::select(annotation_wrapped, probability, annotated)

# Finally, the plot:
ggplot(cellassign_prob_df) +
  aes(x = probability,
      y = annotation_wrapped, 
      fill = annotated) + 
  ggridges::stat_density_ridges(quantile_lines = TRUE,
                                quantiles = 2, 
                                alpha = 0.6) +
  scale_fill_viridis_d() +
  labs(
    x = "CellAssign probability",
    y = "Annotated cell type", 

>     fill = "Cell annotated as given cell type"

Reviewer: Should we make this two lines so that we decrease the space the legend takes up?
